test/mpi/dtpools/src/dtpools_internal.c

```diff
@@ -0,0 +1,1167 @@
+        DTPI_OBJ_CHECK_BUF_AND_JUMP(uint8_t, ptrs.uint8_ptr);
+    }
+    else if (!STRCMP(basic_type_name, "MPI_UINT16_T")) {
+        DTPI_OBJ_CHECK_BUF_AND_JUMP(uint16_t, ptrs.uint16_ptr);
+    }
+    else if (!STRCMP(basic_type_name, "MPI_UINT32_T")) {
+        DTPI_OBJ_CHECK_BUF_AND_JUMP(uint32_t, ptrs.uint32_ptr);
+    }
+    else if (!STRCMP(basic_type_name, "MPI_UINT64_T")) {
+        DTPI_OBJ_CHECK_BUF_AND_JUMP(uint64_t, ptrs.uint64_ptr);
+    }
+
+  fn_exit:
+    return err;
+
+  fn_fail:
+    err = DTP_ERR_OTHER;
+    goto fn_exit;
+}
+
+/* --------------------------------------------------------- */
+/* Datatype Pool Object Create Functions                     */
+/* --------------------------------------------------------- */
+
+int DTPI_Struct_create(struct params *par, DTP_t dtp)
+{
+    int i, err = DTP_SUCCESS;
+    int basic_type_count = 0;
+    int *basic_type_sizes = NULL;
+    int *basic_type_counts;
+    int len, size = 0;
+    MPI_Aint displs, *basic_type_displs = NULL;
+    MPI_Datatype *basic_types;
+    void *buf;
+    DTPI_t *dtpi = NULL;
+    char **basic_type_names = NULL;
+    char type_name[TYPE_NAME_MAXLEN] = { 0 };
+
+    /* allocate space for private datatype info */
+    dtpi = (DTPI_t *) malloc(sizeof(*dtpi));
+    if (!dtpi) {
+        err = DTP_ERR_OTHER;
+        fprintf(stdout, "Out of memory in %s\n", __FUNCTION__);
+        fflush(stdout);
+        goto fn_fail;
+    }
+
+    basic_type_count = dtp->DTP_type_signature.DTP_pool_struct.basic_type_count;
+    basic_type_names = (char **) malloc(sizeof(*basic_type_names) * basic_type_count);
+    basic_type_sizes = (int *) malloc(sizeof(*basic_type_sizes) * basic_type_count);
+    basic_type_displs = (MPI_Aint *) malloc(sizeof(*basic_type_displs) * basic_type_count);
+    if (!basic_type_names || !basic_type_sizes || !basic_type_displs) {
+        err = DTP_ERR_OTHER;
+        fprintf(stdout, "Out of memory in %s\n", __FUNCTION__);
+        fflush(stdout);
+        goto fn_fail;
+    }
+
+    for (i = 0; i < basic_type_count; i++) {
+        basic_type_names[i] = (char *) malloc(sizeof(**basic_type_names) * BASIC_TYPE_NAME_MAXLEN);
+        if (!basic_type_names[i]) {
+            err = DTP_ERR_OTHER;
+            fprintf(stdout, "Out of memory in %s\n", __FUNCTION__);
+            fflush(stdout);
+            goto fn_fail;
+        }
+    }
+
+    /* compute struct size */
+    basic_type_counts = dtp->DTP_type_signature.DTP_pool_struct.DTP_basic_type_count;
+    basic_types = dtp->DTP_type_signature.DTP_pool_struct.DTP_basic_type;
+    for (i = 0; i < basic_type_count; i++) {
+        err = MPI_Type_size(basic_types[i], &basic_type_sizes[i]);
+        if (err) {
+            DTPI_Print_error(err);
+        }
+        err = MPI_Type_get_name(basic_types[i], basic_type_names[i], &len);
+        if (err) {
+            DTPI_Print_error(err);
+        }
+        size += (basic_type_sizes[i] * basic_type_counts[i]);
+    }
+
```

Comment:
The `compute struct size` routine seems unnecessary. You can just call `MPI_Type_get_extent` with the struct type.

Reply:
Ok I will fix this